src/v/io/cache.h

```diff
@@ -0,0 +1,673 @@
+ * using the s3-fifo eviction algorithm is to add a control structure to the
+ * entry that stores internal metadata and provides communication between the
+ * application and the cache.
+ *
+ *     struct entry {
+ *         std::string address;
+ *         std::string description;
+ *         const char* image;
+ *
+ *         io::cache_hook hook;
+ *     };
+ *
+ * Next we'll define an instance of the cache to hold objects of type entry.
+ *
+ *     struct evict {
+ *         bool operator()(entry& e) noexcept {
+ *             e.address.clear();
+ *             e.description.clear();
+ *             free(e.image);
+ *             return true;
+ *         }
+ *     };
+ *
+ *     using cache_type = io::cache<entry, &entry::hook,
+ *       evict, io::default_cache_cost>;
+ *
+ * The `evict` callable implements the application-specific eviction operation,
+ * which in this case frees allocated memory. The type `io::default_cache_cost`
+ * provides a default cost function useful to specifiying cache capacity in
+ * terms of the number of entries in the cache.
+ *
+ * Finally, instantiate the cache with a maximum capacity and the desired size
+ * of the small queue (10% is a reasonable choice).
+ *
+ *     io::cache::config config{.capacity = 250, .small_size = 25};
+ *     cache_type cache{config};
+ *
+ * Call `cache::insert` to add an entry into the cache:
+ *
+ *     auto res = entries.emplace("e0", std::make_unique<entry>(...));
+ *     cache.insert(*res.first.second);
+ *
+ * When the application interacts with an entry it should communicate this to
+ * the cache by touching the entry:
+ *
+ *     auto e = lookup("e0");
+ *     compress(e.image);
+ *     e.hook.touch();
+ *
+ * When inserting an item into the cache, existing items may be evicted to make
+ * space for the new entry. The application can detect evicted entries either by
+ * detecting side effects from the eviction function, or by calling `evicted()`
+ * on the entry's hook member.
+ *
+ * If ane entry has been evicted, it should be re-inserted into the cache after
+ * handling the cache miss. For example:
+ *
+ *     auto it = entries.find("e0");
+ *
+ *     // entry is not in the index
+ *     if (it == entries.end()) {
+ *         auto e0 = handle_cache_miss("e0");
+ *         auto res = entries.emplace("e0", std::move(e0));
+ *         cache.insert(*res.first.second);
+ *     }
+ *
+ *     // entry is in index, but was evicted
+ *     if (it->second.hook.evicted()) {
+ *         handle_cache_miss("e0", it->second);
+ *         cache.insert(*it->second);
+ *     }
+ *
+ * Because the cache may evict random entries, the application should consider
+ * periodically removing evicted items from the index. However, it is generally
+ * desired to not remove items that remain on the ghost queue. The application
+ * can query ghost queue membership using:
+ *
+ *     if (entry.hook.evicted()) {
+ *         if (cache.ghost_queue_contains(entry)) {
+ *             // can fully remove
```

Comment:
Should this be reversed? Wouldn't we only want to fully remove if the ghost queue doesn't contain the entry?